Wikipedia/Code/SearchLanguagesBarViewController.swift

```diff
@@ -0,0 +1,145 @@
+    func searchLanguagesBarViewController(controller: SearchLanguagesBarViewController, didChangeCurrentlySelectedSearchLanguage language: MWKLanguageLink)
+}
+
+class SearchLanguagesBarViewController: UIViewController, WMFPreferredLanguagesViewControllerDelegate {
+    weak var delegate: SearchLanguagesBarViewControllerDelegate?
+    
+    @IBOutlet private var languageButtons: [UIButton] = []
+    @IBOutlet private var otherLanguagesButton: UIButton?
+    @IBOutlet private var heightConstraint: NSLayoutConstraint?
+    
+    private var previousFirstLanguage: MWKLanguageLink?
+    private var hidden: Bool = false {
+        didSet {
+            if(hidden){
+                heightConstraint!.constant = 0
+                view.hidden = true
+            }else{
+                heightConstraint!.constant = 44
+                view.hidden = false
+            }
+        }
+    }
+
+    private var _currentlySelectedSearchLanguage: MWKLanguageLink?
+    private(set) var currentlySelectedSearchLanguage: MWKLanguageLink? {
+        get {
+            return _currentlySelectedSearchLanguage
+        }
+        set {
+            _currentlySelectedSearchLanguage = newValue
+            NSUserDefaults.wmf_userDefaults().wmf_setCurrentSearchLanguageDomain(newValue?.siteURL())
+            delegate?.searchLanguagesBarViewController(self, didChangeCurrentlySelectedSearchLanguage: newValue!)
+            updateLanguageBarLanguageButtons()
+        }
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        _ = languageButtons.map{$0.tintColor = UIColor.wmf_blueTintColor()}
+        otherLanguagesButton?.setBackgroundImage(UIImage.wmf_imageFromColor(UIColor.whiteColor()), forState: .Normal)
+        otherLanguagesButton?.setBackgroundImage(UIImage.wmf_imageFromColor(UIColor(white: 0.9, alpha: 1.0)), forState: .Highlighted)
+        otherLanguagesButton?.setTitle(localizedStringForKeyFallingBackOnEnglish("main-menu-title"), forState: .Normal)
+        otherLanguagesButton?.titleLabel?.font = UIFont.wmf_subtitle()
+
+        _currentlySelectedSearchLanguage = selectedLanguage()
+        assert(_currentlySelectedSearchLanguage != nil, "Expected language not set")
+        
+        previousFirstLanguage = languageBarLanguages().first
+    }
+    
+    override func viewWillAppear(animated: Bool) {
+        super.viewWillAppear(animated)
+        updateLanguageBarLanguageButtons()
+        hidden = !NSUserDefaults.wmf_userDefaults().wmf_showSearchLanguageBar()
+    }
+
+    override func viewDidAppear(animated: Bool) {
+        super.viewDidAppear(animated)
+        selectFirstLanguageIfNoneSelectedOrIfFirstLanguageHasChanged()
+        previousFirstLanguage = languageBarLanguages().first
+    }
+    
+    private func languageBarLanguages() -> [MWKLanguageLink] {
+        return MWKLanguageLinkController.sharedInstance().preferredLanguages.enumerate().flatMap{ $0.index < 3 ? $0.element : nil }
```

Comment:
`preferredLanguages.prefix(3)` will grab the first 3 elements without flatmapping and is bounds safe